cpp/examples/arrow/dataset-parquet-scan-example.cc

```diff
@@ -0,0 +1,137 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <cstdlib>
+#include <iostream>
+
+#include <arrow/api.h>
+
+#include <arrow/dataset/dataset.h>
+#include <arrow/dataset/discovery.h>
+#include <arrow/dataset/file_base.h>
+#include <arrow/dataset/file_parquet.h>
+#include <arrow/dataset/filter.h>
+#include <arrow/dataset/scanner.h>
+#include <arrow/filesystem/localfs.h>
+#include <arrow/util/iterator.h>
+#include <arrow/util/task_group.h>
+#include <arrow/util/thread_pool.h>
+
+using arrow::field;
+using arrow::int16;
+using arrow::Schema;
+using arrow::Table;
+
+using arrow::internal::TaskGroup;
+using arrow::internal::ThreadPool;
+
+using arrow::fs::FileSystem;
+using arrow::fs::LocalFileSystem;
+using arrow::fs::Selector;
+
+using arrow::dataset::Dataset;
+using arrow::dataset::DataSource;
+using arrow::dataset::DataSourceDiscovery;
+using arrow::dataset::DataSourceVector;
+using arrow::dataset::Expression;
+using arrow::dataset::FieldExpression;
+using arrow::dataset::FileFormat;
+using arrow::dataset::FileSystemDataSourceDiscovery;
+using arrow::dataset::ParquetFileFormat;
+using arrow::dataset::PartitionScheme;
+using arrow::dataset::Scanner;
+using arrow::dataset::ScannerBuilder;
+using arrow::dataset::ScanTask;
+using arrow::dataset::string_literals::operator"" _;
+
+#define ABORT_ON_FAILURE(expr)                     \
+  do {                                             \
+    arrow::Status status_ = (expr);                \
+    if (!status_.ok()) {                           \
+      std::cerr << status_.message() << std::endl; \
+      abort();                                     \
+    }                                              \
+  } while (0);
+
+std::shared_ptr<Dataset> GetDatasetFromPath(FileSystem* fs,
+                                            std::shared_ptr<FileFormat> format,
+                                            std::string path) {
+  // Find all files under `path`
```

Comment:
Could you describe somewhere what the files under path should look like? And what the paths should be, if that's relevant for partition discovery?

Reply:
This example doesn't use partitions yet. I didn't want to add it because it would achieve nothing.